rust-runtime/aws-smithy-mocks/src/rule.rs

```diff
@@ -0,0 +1,336 @@
+    {
+        self.sequence().http_response(response_fn).build()
+    }
+}
+
+type SequenceGeneratorFn<O, E> = Arc<dyn Fn() -> MockResponse<O, E> + Send + Sync>;
+
+/// A builder for creating response sequences
+pub struct ResponseSequenceBuilder<I, O, E> {
+    /// The response generators in the sequence
+    generators: Vec<SequenceGeneratorFn<O, E>>,
+
+    /// Function that determines if this rule matches a request
+    input_filter: MatchFn,
+
+    /// Marker for the input, output, and error types
+    _marker: std::marker::PhantomData<I>,
+}
+
+impl<I, O, E> ResponseSequenceBuilder<I, O, E>
+where
+    I: fmt::Debug + Send + Sync + 'static,
+    O: fmt::Debug + Send + Sync + 'static,
+    E: fmt::Debug + Send + Sync + std::error::Error + 'static,
+{
+    /// Create a new response sequence builder
+    pub(crate) fn new(input_filter: MatchFn) -> Self {
+        Self {
+            generators: Vec::new(),
+            input_filter,
+            _marker: std::marker::PhantomData,
+        }
+    }
+
+    /// Add a modeled output response to the sequence
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// let rule = mock!(Client::get_object)
+    ///     .sequence()
+    ///     .output(|| GetObjectOutput::builder().build())
+    ///     .build();
+    /// ```
+    pub fn output<F>(mut self, output_fn: F) -> Self
+    where
+        F: Fn() -> O + Send + Sync + 'static,
+    {
+        let generator = Arc::new(move || MockResponse::Output(output_fn()));
+        self.generators.push(generator);
+        self
+    }
+
+    /// Add a modeled error response to the sequence
+    pub fn error<F>(mut self, error_fn: F) -> Self
+    where
+        F: Fn() -> E + Send + Sync + 'static,
+    {
+        let generator = Arc::new(move || MockResponse::Error(error_fn()));
+        self.generators.push(generator);
+        self
+    }
+
+    /// Add an HTTP status code response to the sequence
+    pub fn http_status(mut self, status: u16, body: Option<String>) -> Self {
+        let status_code = StatusCode::try_from(status).unwrap();
+
+        let generator: SequenceGeneratorFn<O, E> = match body {
+            Some(body) => Arc::new(move || {
+                MockResponse::Http(HttpResponse::new(status_code, SdkBody::from(body.clone())))
+            }),
+            None => Arc::new(move || {
+                MockResponse::Http(HttpResponse::new(status_code, SdkBody::empty()))
+            }),
+        };
+
+        self.generators.push(generator);
+        self
+    }
+
+    /// Add an HTTP response to the sequence
+    pub fn http_response<F>(mut self, response_fn: F) -> Self
+    where
+        F: Fn() -> HttpResponse + Send + Sync + 'static,
+    {
+        let generator = Arc::new(move || MockResponse::Http(response_fn()));
+        self.generators.push(generator);
+        self
+    }
+
+    /// Repeat the last added response multiple times (total count)
+    pub fn times(mut self, count: usize) -> Self {
```

Comment:
What would happen if we specified `.times(0)`? I've worked with google mocks for C++ where `TIMES(0)` means that no call is expected for the mock method.

Reply:
Right now anything less than 1 has no additional effect.

Reply:
Got it, we should probably call out in the doc stating anything less than 1 has no effect or leave `TODO`.